### PR TITLE
Check datatype of *_Expr classes [closes issue #284]

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for first review from @arporter

**Comment by rupertford** [comment]
Also closes #276 I think.

File: src/fparser/two/Fortran2003.py [hunk 1]
```diff
@@ -4909,39 +4909,193 @@ class Defined_Binary_Op(Base):  # pylint: disable=invalid-name
...
+    subclass_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for a character expression.
+
+        :param str string: Fortran code to check for a match.
+        :return: `None` if there is no match, or a tuple containing \
+                 the matched expression.
+        :rtype: NoneType or (???)
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:4972]: Please complete.

File: src/fparser/two/Fortran2003.py [hunk 2]
```diff
@@ -4909,39 +4909,193 @@ class Defined_Binary_Op(Base):  # pylint: disable=invalid-name
...
+    subclass_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for a default character expression.
+
+        :param str string: Fortran code to check for a match.
+        :return: `None` if there is no match, or a tuple containing \
+                 the matched expression.
+        :rtype: NoneType or (???)
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:5009]: Please complete.

File: src/fparser/two/Fortran2003.py [hunk 3]
```diff
@@ -4909,39 +4909,193 @@ class Defined_Binary_Op(Base):  # pylint: disable=invalid-name
...
+    subclass_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for an integer expression.
+
+        :param str string: Fortran code to check for a match.
+        :return: `None` if there is no match, or a tuple containing \
+                 the matched expression.
+        :rtype: NoneType or (???)
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:5046]: Please complete.

File: src/fparser/two/Fortran2003.py [hunk 4]
```diff
@@ -4909,39 +4909,193 @@ class Defined_Binary_Op(Base):  # pylint: disable=invalid-name
...
+    subclass_names = []
+
+    @staticmethod
+    def match(string):
+        '''Implements the matching for a numeric expression.
+
+        :param str string: Fortran code to check for a match.
+        :return: `None` if there is no match, or a tuple containing \
+                 the matched expression.
+        :rtype: NoneType or (???)
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:5083]: Please complete and throughout.

File: src/fparser/two/tests/fortran2003/test_primary_r701.py [hunk 5]
```diff
@@ -149,37 +149,38 @@ def test_array_constructor():
         expected_str='[1.2, 2.3E+2, - 5.1E-3]')
 
 
-@pytest.mark.xfail(reason="Matches Designator before Structure Constructor "
-                   "(issue #284)")
 @pytest.mark.usefixtures("f2003_create")
```
Comment (arporter) [inline src/fparser/two/tests/fortran2003/test_primary_r701.py:152]: There's a test in `test_fortran2003.py` that refers to R701.h. Perhaps move it to this file?

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good. It's nice to be able to tighten things up, even if only a bit.
Just minor things to address.
I've checked all files for pylint/pycodestyle.

File: src/fparser/two/tests/fortran2003/test_primary_r701.py (hunk 5 above)
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_primary_r701.py:152]: OK, done.

File: src/fparser/two/Fortran2003.py [hunk 6]
```diff
@@ -4909,39 +4909,193 @@ class Defined_Binary_Op(Base):  # pylint: disable=invalid-name
...
 
-class Int_Expr(Base):  # R727
-    """
-    <int-expr> = <expr>
-    """
-    subclass_names = ['Expr']
+        :param str string: Fortran code to check for a match.
+        :return: `None` if there is no match, or a tuple containing \
+                 the matched expression.
+        :rtype: NoneType or (???)
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2003.py:4935]: oops

File: src/fparser/two/Fortran2003.py (hunk 1 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py:4972]: oops

File: src/fparser/two/Fortran2003.py (hunk 2 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py:5009]: oops

File: src/fparser/two/Fortran2003.py (hunk 3 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py:5046]: oops

File: src/fparser/two/Fortran2003.py (hunk 4 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py:5083]: oops. Sorry, I forgot to add this then replicated with cut-n-paste.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Latest changes are pylint/pycodestyle clean.
Coverage of the diff is still 100%.
No need to update docs.

**Pull request merged**